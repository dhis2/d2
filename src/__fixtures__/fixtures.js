export default (function x() {
    const fixtures = {};

    function getFixture(fixtureName) {
        if (fixtures && fixtures[fixtureName]) {
            // Make sure we return a new object
            return JSON.parse(JSON.stringify(fixtures[fixtureName]));
        }
        throw new Error(['Fixture', fixtureName, 'does not exist'].join(' '));
    }

    function addFixture(fixtureName, fixtureData) {
        if (fixtureName && fixtureData && typeof fixtureName === 'string') {
            fixtures[fixtureName] = fixtureData;
        }
    }

    addFixture('/api/attributes', require('./json/api/schemas/attributes.json'));
    addFixture('/api/dataElements', require('./json/api/dataElements.json'));
    addFixture('/api/organisationUnits', require('./json/api/organisationUnits.json'));
    addFixture('/api/schemas', require('./json/api/schemas.json'));
    addFixture('/api/schemas/dataElement', require('./json/api/schemas/dataElement.json'));
    addFixture('/api/schemas/dataSet', require('./json/api/schemas/dataSet'));
    addFixture('/api/schemas/indicatorGroup', require('./json/api/schemas/indicatorGroup.json'));
    addFixture('/api/schemas/organisationUnit', require('./json/api/schemas/organisationUnit.json'));
    addFixture('/api/schemas/organisationUnitGroupSet', require('./json/api/schemas/organisationUnitGroupSet'));
    addFixture('/api/schemas/validationRule', require('./json/api/schemas/validationRule'));
    addFixture('/api/schemas/program', require('./json/api/schemas/program'));
    addFixture('/api/schemas/programNotificationTemplate', require('./json/api/schemas/programNotificationTemplate'));
    addFixture('/api/schemas/user', require('./json/api/schemas/user.json'));
    addFixture('/api/schemas/legendSet', require('./json/api/schemas/legendSet.json'));
    addFixture('/api/schemas/legend', require('./json/api/schemas/legend.json'));
    addFixture('/api/userSettings', require('./json/api/userSettings'));
    addFixture('/appStore', require('./json/api/appStore.json'));
    addFixture('/dataElementAttributes', require('./json/dataElementAttributes.json'));
    addFixture('/me', require('./json/api/me.json'));
    addFixture('/me/authorities', require('./json/api/me/authorities.json'));
    addFixture('/modelDefinitions/dataElement', require('./json/modelDefinitions/dataElement.json'));
    addFixture('/singleUserAllFields', require('./json/singleUserAllFields.json'));
    addFixture('/singleUserOwnerFields', require('./json/singleUserOwnerFields.json'));
    addFixture('/api/legendSets/k1JHPfXsJND', require('./json/api/legendSets/k1JHPfXsJND.json'));
    addFixture('/api/analytics/dataValueSet', require('./json/api/analytics/dataValueSet.json'));
    addFixture('/api/analytics/rawData', require('./json/api/analytics/rawData.json'));
    addFixture('/api/analytics/aggregate', require('./json/api/analytics/aggregate.json'));
    addFixture('/api/analytics/count', require('./json/api/analytics/count.json'));
    addFixture('/api/analytics/query', require('./json/api/analytics/query.json'));
    addFixture('/api/analytics/cluster', require('./json/api/analytics/cluster.json'));

    return {
        get: getFixture,
        add: addFixture,
    };
}());
