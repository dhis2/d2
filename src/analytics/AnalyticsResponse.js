import AnalyticsResponseHeader from './AnalyticsResponseHeader'

const booleanMap = {
    0: 'No', // XXX i18n.no || 'No',
    1: 'Yes', // i18n.yes || 'Yes',
}

const OUNAME = 'ouname'
const OU = 'ou'

const DEFAULT_COLLECT_IGNORE_HEADERS = [
    'psi',
    'ps',
    'eventdate',
    'longitude',
    'latitude',
    'ouname',
    'oucode',
    'eventdate',
    'eventdate',
]

const DEFAULT_PREFIX_IGNORE_HEADERS = ['dy', ...DEFAULT_COLLECT_IGNORE_HEADERS]

const getParseMiddleware = type => {
    switch (type) {
        case 'STRING':
        case 'TEXT':
            return value => `${value}`
        case 'INTEGER':
        case 'NUMBER':
            return value =>
                !Number.isNaN(+value) && Number.isFinite(+value)
                    ? parseFloat(+value)
                    : value
        default:
            return value => value
    }
}

const isPrefixHeader = (header, dimensions) => {
    if (DEFAULT_PREFIX_IGNORE_HEADERS.includes(header.name)) {
        return false
    }

    return Boolean(Array.isArray(dimensions) && dimensions.length === 0)
}

const isCollectHeader = (header, dimensions) => {
    if (DEFAULT_COLLECT_IGNORE_HEADERS.includes(header.name)) {
        return false
    }

    return Boolean(Array.isArray(dimensions) && dimensions.length === 0)
}

const getPrefixedId = (id, prefix) => `${prefix || ''} ${id}`

const getNameByIdsByValueType = (id, valueType) => {
    if (valueType === 'BOOLEAN') {
        return booleanMap[id]
    }

    return id
}

class AnalyticsResponse {
    constructor(response) {
        if (response) {
            this.response = response
            this.headers = this.extractHeaders()
            this.rows = this.extractRows()
            this.metaData = this.extractMetadata()
        }
    }

    extractHeaders() {
        const { dimensions } = this.response.metaData
        const headers = this.response.headers || []

        return headers.map(
            (header, index) =>
                new AnalyticsResponseHeader(header, {
                    isPrefix: isPrefixHeader(header, dimensions[header.name]),
                    isCollect: isCollectHeader(header, dimensions[header.name]),
                    index,
                })
        )
    }

    extractRows() {
        const headersWithOptionSet = this.headers.filter(
            header => header.optionSet
        )
        let { rows } = this.response

        if (headersWithOptionSet.length) {
            rows = rows.slice()

            const optionCodeIdMap = this.optionCodeIdMap()

            // replace option code with option uid
            headersWithOptionSet.forEach(header => {
                rows.forEach((row, index) => {
                    const id = optionCodeIdMap[header.name][row[header.index]]

                    if (id) {
                        rows[index][header.index] = id
                    }
                })
            })
        }

        return rows
    }

    extractMetadata() {
        const metaData = { ...this.response.metaData }

        const { dimensions, items } = metaData

        // populate metaData dimensions and items
        this.headers
            .filter(
                header => !DEFAULT_COLLECT_IGNORE_HEADERS.includes(header.name)
            )
            .forEach(header => {
                let ids

                // collect row values
                if (header.isCollect) {
                    ids = this.getSortedUniqueRowIdStringsByHeader(header)
                    dimensions[header.name] = ids
                } else {
                    ids = dimensions[header.name]
                }

                if (header.isPrefix) {
                    // create prefixed dimensions array
                    dimensions[header.name] = ids.map(id =>
                        getPrefixedId(id, header.name)
                    )

                    // create items
                    dimensions[header.name].forEach((prefixedId, index) => {
                        const id = ids[index]
                        const { valueType } = header

                        const name = getNameByIdsByValueType(id, valueType)

                        items[prefixedId] = { name }
                    })
                }
            })

        // for events, add items from 'ouname'
        if (this.hasHeader(OUNAME) && this.hasHeader(OU)) {
            const ouNameHeaderIndex = this.getHeader(OUNAME).getIndex()
            const ouHeaderIndex = this.getHeader(OU).getIndex()
            let ouId
            let ouName

            this.rows.forEach(row => {
                ouId = row[ouHeaderIndex]

                if (items[ouId] === undefined) {
                    ouName = row[ouNameHeaderIndex]

                    items[ouId] = {
                        name: ouName,
                    }
                }
            })
        }

        return metaData
    }

    getHeader(name) {
        return this.headers.find(header => header.name === name)
    }

    hasHeader(name) {
        return this.getHeader(name) !== undefined
    }

    getSortedUniqueRowIdStringsByHeader(header) {
        const parseByType = getParseMiddleware(header.valueType)
        const parseString = getParseMiddleware('STRING')

        const rowIds = Array.from(
            // unique values
            new Set(
                this.rows.map(responseRow =>
                    parseByType(responseRow[header.index])
                )
            )
            // remove empty values
        ).filter(id => id !== '')

        return rowIds.sort().map(id => parseString(id))
    }

    optionCodeIdMap() {
        const { dimensions, items } = this.response.metaData
        const map = {}

        this.headers
            .filter(header => typeof header.optionSet === 'string')
            .forEach(header => {
                const optionIds = dimensions[header.name]

                map[header.name] = optionIds
                    .map(id => ({
                        [items[id].code]: id,
                    }))
                    .reduce((acc, obj) => Object.assign(acc, obj), {})
            })

        return map
    }

    sortOrganisationUnitsHierarchy() {
        const organisationUnits = this.metaData.dimensions.ou

        organisationUnits.forEach((organisationUnit, i) => {
            const hierarchyPrefix = this.metaData.ouHierarchy[organisationUnit]
            const hierarchyIds = [organisationUnit]
            const hierarchyNames = []

            hierarchyPrefix
                .split('/')
                .reverse()
                .forEach(ouId => {
                    hierarchyIds.unshift(ouId)
                })

            hierarchyIds.forEach(ouId => {
                if (this.metaData.items[ouId]) {
                    hierarchyNames.push(this.metaData.items[ouId].name)
                }
            })

            organisationUnits[i] = {
                id: organisationUnit,
                fullName: hierarchyNames.join(' / '),
            }
        })

        // XXX how does this work with different languages/collations?
        organisationUnits.sort((a, b) => {
            const aFullName = a.fullName
            const bFullName = b.fullName

            if (aFullName < bFullName) {
                return -1
            }

            return aFullName > bFullName ? 1 : 0
        })

        this.metaData.dimensions.ou = organisationUnits.map(ou => ou.id)
    }
}

export default AnalyticsResponse
