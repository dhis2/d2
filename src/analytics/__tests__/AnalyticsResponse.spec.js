import fixtures from '../../__fixtures__/fixtures'
import AnalyticsResponse from '../AnalyticsResponse'

const fixture = fixtures.get('/api/analytics/response')

let response

describe('AnalyticsResponse', () => {
    beforeEach(() => {
        response = new AnalyticsResponse(fixture)
    })

    describe('constructor', () => {
        it('should not be allowed to be called without new', () => {
            expect(() => AnalyticsResponse()).toThrowErrorMatchingSnapshot()
        })

        it('should set the response when passed as argument', () => {
            expect(response.response).toEqual(fixture)
        })
    })

    describe('properties', () => {
        describe('.extractHeaders()', () => {
            it('should return the headers from the response', () => {
                const headers = response.extractHeaders()

                expect(headers).toHaveLength(4)
            })
        })

        describe('.extractRows()', () => {
            it('should return the rows from the response', () => {
                const rows = response.extractRows()

                expect(rows).toHaveLength(312)
            })
        })

        describe('.extractMetadata()', () => {
            it('should return the metadata from the response', () => {
                const metadata = response.extractMetadata()

                expect(metadata.dimensions).toEqual(fixture.metaData.dimensions)
                expect(metadata.items).toEqual(fixture.metaData.items)
            })

            // TODO
            // test with OUNAME case
        })
    })
})
