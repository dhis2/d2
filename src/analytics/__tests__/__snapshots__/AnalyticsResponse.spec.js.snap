// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsResponse constructor should not be allowed to be called without new 1`] = `"Class constructor AnalyticsResponse cannot be invoked without 'new'"`;
